### PR TITLE
WIP: Implement opencensus for tracing

### LINKED ISSUE #673: Enable opencensus on the agent
Here are some general questions that I have about the current tracing implementation and the one that it will be implemented
1. I understand that we will use the host.go from (https://github.com/mcastelino/custom_exporter/) at the host and that I need to modify the tracing.go (to include part of the vm.go code from Manohar's). The set of changes involve to remove the use of the "github.com/uber/jaeger-client-go/config", removing the use of the `jaeger-client-socat-redirector.service`,
`kata-redirect-agent-output-to-journal.conf` and `kata-journald-host-redirect.service` as they are use as a way to communicate Jaeger client code with the jaeger agent on the host (vsock limitation). Are these asumptions correct? Now this also will have an impact on the Makefile as they are part of it.
2. Is there a way that I can test the changes that I am doing at the tracing.go? I tried to build it alone and it seems that is not possible as you need to compile or build everything.
3. Currently, I submitted some PRs https://github.com/kata-containers/agent/pull/671 and https://github.com/kata-containers/agent/commit/3e625657823249a7ffdb35d5d04f573d61053c43 in order to make easier the changes of the name of the functions as I see on the opencensus API the way that we for example set a tag is different for the opentracing stuff.

### PR DESCRIPTION
**Pull request by GabyCT**
This implements opencensus for tracing on the agent.

Fixes #673

Signed-off-by: Gabriela Cervantes <gabriela.cervantes.tellez@intel.com>

### CODE REVIEW

**Comment by GabyCT** [comment]
This is a WIP, currently I am focusing in start to modify the `tracing.go` using `opencensus`. The first part that I did was to modify the functions of `start the span`, `start the span from context` (by using opencensus definition). Also, I modify the way that we perform a `tag`, `finish a span`, etc. One of the issues that I am seeing is that in `opentracing` there is a function like `StartSpanFromContext`, which is non-existing on `opencensus` so I am not quite sure how this will be handled.
I removed the `createTracer` function because this is use to setup the `tracer` on `opentracing`, however, looking the info at https://opencensus.io/quickstart/go/tracing/, that does not exist on `opencensus`.
One of the issues that I am seeing which will impact the way that things are done on tracing is that
in the case of `opentracing` we needed to setup a tracer, start a trace, create a child span, etc. However, on opencensus (https://opencensus.io/quickstart/go/tracing/), we need to configure the exporter, a sampler, create a span, create a child span, create an annotation, etc. 
This will have a considerable impact on functions that are defined in several files like https://github.com/kata-containers/agent/pull/681/files#diff 376d028b7ce6b85c76a370dd6d94ece7L127
and also like https://github.com/kata-containers/agent/pull/681/files#diff-376d028b7ce6b85c76a370dd6d94ece7L177 as I think they
will need to be modified, wdyt? 
/cc @egernst @jodh-intel @amshinde

**Comment by GabyCT** [comment]
I already start to merging the code from Manohar's with our current implementation, I also modified the `tracing.go` to use opencensus, please take a look

**Comment by GabyCT** [comment]
I need to modify https://github.com/kata-containers/agent/blob/master/agent.go#L1113, as it is using something attached to opentracing and it looks that is the grpc implementation so I need to take a look for grpc on opencensus

File: tracing.go [hunk 1]
```diff
@@ -6,176 +6,273 @@
...
+// spanNewContext returns a new context with the givien span attached
+func spanNewContext(parent context.Context, s *agentSpan) context.Context {
+	return trace.NewContext(parent, s.span)
 }
 
-type agentTracer struct {
-	tracer opentracing.Tracer
+// setTag sets the name of the span, if it is recording events
+func (a *agentSpan) setTag(name string, value interface{}) {
+	a.span.SetName(fmt.Sprintf("%s=%v" ,name, value))
```
Comment (jodh-intel) [inline tracing.go]: This doesn't look correct to me. According to the API, `SetName()` sets the name of the trace itself. But tags in openTracing are "name=value" pairs. Each trace can have zero or more tags. I think we should be using https://godoc.org/go.opencensus.io/tag?

Comment (GabyCT) [reply tracing.go]: @jodh-intel , looking the API for tag on opencensus, I can't find a function where we have as input(key, value) like we have in opentracing, if you see this
```
func MustNewKey(name string) Key
func NewKey(name string) (Key, error)
func (k Key) Name() string
```
they want as an input a `name` and they return a `key` and `error` which looks very similar to the current implementation that I have, wdyt?

Comment (GabyCT) [reply tracing.go]: @jodh-intel , I change the tag implementation, thanks

File: tracing.go [hunk 2]
```diff
@@ -6,176 +6,273 @@
...
+	span *trace.Span
 }
 
 // The first trace span
 var rootSpan *agentSpan
 
-// Implements jaeger-client-go.Logger interface
-type traceLogger struct {
+// startSpan starts a new child span of the current span in the context
+func startSpan(ctx context.Context, name string) (context.Context, *agentSpan) \
```
[the remark this answers is not in the document]
Comment (GabyCT) [reply tracing.go]: done

File: tracing.go [hunk 3]
```diff
@@ -6,176 +6,273 @@
...
+func spanFromContext(ctx context.Context) *agentSpan {
+	var a agentSpan
+	a.span = trace.FromContext(ctx)
+	return &a
 }
 
-func (a *agentSpan) finish() {
-	a.span.Finish()
+// spanStartSpanWithRemoveParent starts a new child span of the span from the given parent
+func spanStartSpanWithRemoteParent(ctx context.Context, name string, parent tra\
```
[the remark this answers is not in the document]
Comment (GabyCT) [reply tracing.go]: done

File: tracing.go [hunk 4]
```diff
@@ -6,176 +6,273 @@
...
-	// Create the root span (which is .Finish()'d by stopTracing())
-	span := tracer.startSpan(rootSpanName)
+	// Create the rootspan
+	ctx, span := startSpan(ctx, rootSpanName)
 	span.setTag("source", "agent")
 	span.setTag("root-span", "true")
 
-	// See comment in trace().
+	// Annotate our span to capture metadata about our operation
+	span.spanAnnotate([]trace.Attribute{trace.Int64Attribute("invocations", 1)}, "Invoked it")
```
[the remark this answers is not in the document]
Comment (GabyCT) [reply tracing.go]: @jodh-intel , I remove this part, I have been looking opencensus examples and I do not think we need this

**Comment by chavafg** [comment]
>I need to modify https://github.com/kata-containers/agent/blob/master/agent.go#L1113, as it is using something attached to opentracing (opentracing.Tracer) that has not an equivalent in opencensus, @egernst what do you think of this?
That Tracer looks that is available in opentelemtry https://godoc.org/go.opentelemetry.io/api/trace#Span, so @egernst which is the approach that we are going to take on this?

@egernst can you give an input here? afaiu @GabyCT is waiting for consensus on how we should move on.

**Comment by GabyCT** [comment]
It looks like an alpha version of some components for opentelemetry has been released https://github.com/open-telemetry/opentelemetry-go
/cc @jodh-intel @egernst

**Comment by egernst** [comment]
If opentelemetry is using Tracer, it may make sense to give this a try. Are you able to do any functional tests?

**Comment by grahamwhaley** [comment]
@GabyCT @jodh-intel - do we think this is dead yet, or ever likely to be revived?

**Comment by jodh-intel** [comment]
Let's wait for @GabyCT's confirmation, but as far as I'm concerned, we no longer need this due to https://github.com/kata-containers/kata-containers/issues/60.

**Pull request closed without merging**